### DISCUSSION TITLE
RFC 7523 JWTs as Authorization Grants

### DISCUSSION THREAD

**Discussion by jbman** [keycloak/keycloak#14823] [Ideas]
## Summary
Keycloak supports ["Using JWTs for Client Authentication"](https://www.rfc-editor.org/rfc/rfc7523#section-2.2) from RFC 7523.
The same RFC also defines ["Using JWTs as Authorization Grants"](https://www.rfc-editor.org/rfc/rfc7523#section-2.1):

>    This document defines how a JWT Bearer Token can be used to request
   an access token when a client wishes to utilize an existing trust
   relationship, expressed through the semantics of the JWT, without a
   direct user-approval step at the authorization server.
_from [RFC 7523 - Section 1](https://tools.ietf.org/html/rfc7523#section-1)_

When this is supported, the client can use grant_type ``urn:ietf:params:oauth:grant-type:jwt-bearer``. The client sends a JWT token as ``assertion`` and may request a specific OAuth scope. The JWT token contains the ``sub`` claim which specifies the user for the resulting access token.

In this discussion we would like to receive feedback on the idea to support this grant at Keycloak and also on the options how it could be supported.

## Option 1: Generic support

Use case: **Trust a client application to handle user authentication**

The grant is useful if user authentication should be handled completely by a client instead of the authorization server. The client can issue the tokens on its own and get an access token for any user.

Possible implementation at Keycloak:

Support at Keycloak requires configuration of a trusted issuer and its public key or public Key JWKS endpoint. The JWT bearer grant would be enabled at a client by configuring a set of trusted issuer configurations

Benefits:
* Implementation could be fully compliant with RFC, but has additional restriction of client authentication.

## Option 2: Support for OIDC ID tokens of configured Identity providers

Use case: **Allow silent authorization by accepting OpenId Connect ID tokens**

In this use case a client wants to use a valid OpenId Connect ID token to prove user authentication, without the need of any user interaction like login or consent. 

The client can not just get an access token for any use. It needs to obtain an ID token through regular OIDC login.

Possible implementation at Keycloak:

Keycloak already has OIDC IdP configurations for identity brokering. The JWT bearer grant would be enabled at a client by configuring a set of Keycloak OIDC IdPs. Such a client can then use a valid IdP ID token frokm one of these IdPs to get an access token for a user.

Benefits:
* Minimal configuration at Keycloak needed.
* Usage of grant restricted to configured IdPs, so that it is not required to trust the client to have a secure authentication process but the client is only trusted in secure ID token handling.

**Comment by jbman** [comment]
## Other authorization servers supporting the grant

We could find the following servers supporting the grant:

* Ory Hydra –  This documentation defines configuration and constraints: https://www.ory.sh/docs/hydra/guides/oauth2-grant-type-jwt-bearer#oauth2-jwt-bearer-grant-type-validation
* Authlete – Authlete does only receive the request and does generic request validation. Actual authorization is added with custom implementation: https://www.authlete.com/developers/jwt_authorization_grant/
* PingFederate Server - Supports the grant but unclear how trust is configured: https://docs.pingidentity.com/bundle/pingfederate-111/page/klo1564002964192.html

**Comment by jbman** [comment]
Any feedback on that? 

Here is another summary what it is for:
"RFC 7523 JWTs as Authorization Grants" basically allows a client to impersonate any user and get an access token. One specific client is trusted to handle authentication completely on its own.
Existing resource servers usually expect the resource owner as subject in the JWT. A service account client is not enough because it can't a current user into the token.

Here is the case how we would use it:
We have multiple parties who use Keycloak as authorization server. They want to work together as partners and want to provide one application to the user, where all the resources show up. The app could use the OAuth 2.0 auth code flow at each authorization server to get the users resources, but this means the user is directed to the login of each authorization server. 
The goal is to get that working without user interaction. The partners want to trust the application to handle login. The application can go to every authentication server and issue an access token as defined by "RFC 7523 JWTs as Authorization Grants" from machine to machine.

**Comment by ben95cd** [comment]
I work for MITRE and one of my projects has the need to utilize JWTs as authorization grants much like the use case described above. We have a reference implementation for this feature that we'd like to push up to the community.
 
The implementation matches option 1 above where we trust an application to handle user authentication and is compliant with RFC 7523.
 
**Implementation Summary**
 
The implementation includes 3 major changes:
 
1. Add a configuration block in the client Advanced tab that enables administrators to configure trusted issuer, audience, and certificate combinations.
 
This configuration will be used to validate the signed JWTs sent to the token endpoint.
 
2. Add a handler for the jwt-bearer grant type on the token endpoint, as described in RFC-7523.
 
When a client specified jwt-bearer in the grant_type claim of the token request, Keycloak will load a provider from the newly created assertion grant SPI.
 
3. Default assertion grant provider for the assertion grant SPI
 
A default assertion grant provider was implemented. This provider verifies the JWT as described in RFC-7523. First, it confirms the client is confidential. It will then verify the provided JWT by looping through each configured trusted issuer for the client and attempt to verify the JWT using the certificate, checking that the audience and issuer also match the configuration. If the JWT successfully verifies, Keycloak then checks if the client has the 'impersonate' permission for the user specified in the JWT 'sub' claim. Assuming all checks pass, Keycloak creates a session for the user and returns the generated tokens.
 

In this implementation the existing 'impersonate' permission is utilized to determine if the client is authorized to perform this grant type.

**Comment by pedroigor** [comment]
@ben95cd Can't we solve this better by building it on top of the token exchange?

The mechanics in 7523 look kinda similar to how 8693 works. In fact, the SPI you are proposing is something we need to enable in token exchange so that people can extend the feature without having to base their implementation on top of the default token exchange provider.  

The `impersonate` permission you are referring to is also available when doing token exchanges already. It also addresses client authentication in addition to authorization grant (`sub` in https://www.rfc-editor.org/rfc/rfc7523#section-3).

I don't know but there is an overlap between the two. For instance, both are based on 7521. Being 8693 a more recent standard, don't you think it supersedes 7523?

**Reply by ben95cd** [reply]
Hey @pedroigor I appreciate the feedback! 

In my implementation I actually did convert the TokenExchangeContext into a more generic abstract class representing an OIDC token request context, and then extend that to create the AssertionGrantContext. 

I didn't want to refactor too much of the original TokenExchange code but if we wanted to build it off token exchange we could convert the TokenExchangeSpi into a generic OIDCTokenRequestSpi where the supports() function would take the requested grant type as part of the context, and determine if its supported. The handler for the token endpoint [TokenEndpoint.java](https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/protocol/oidc/endpoints/TokenEndpoint.java#L226) could then have a default case then would find the first OIDCTokenRequestProvider supporting that grant type and process the request. 

Is that something that would be in scope for this PR? Or should I submit this PR with the assertion grant changes and then start a new issue to clean up the token exchanger and assertion grant code into a unified Spi?

**Reply by pedroigor** [reply]
@ben95cd Perhaps we can start with what you have and then figure out how it overlaps with the token exchange.

I'm not really thinking about unifying both SPI but include the set of interfaces you are using to handle different assertion types as part of the Token Exchange SPI.

Wdyt?

**Reply by pedroigor** [reply]
Btw, this feature is definitely useful as you can see from the comments you are getting.

Once you send the PR or link to a branch, I can start writing a design document based on your requirements as well as give others the chance to provide more information about their use cases and requirements.

**Reply by ben95cd** [reply]
@pedroigor that sounds great! I'm just working on cleaning up some final changes and then will open up a PR.

**Reply by ben95cd** [reply]
@pedroigor just wanted to let you know that i opened up the PR. https://github.com/keycloak/keycloak/pull/24512

**Comment by schlomo** [comment]
I'm trying to implement https://circleci.com/docs/openid-connect-tokens/ with Keycloak, so that code running in a CI pipeline can exchange a CircleCI token for a JWT impersonating a user or for an auto-generated user matching the metadata in the CircleCI token (which has all the info on git repo, branch etc.).

Can this proposal solve that use case too? Or what would be right way to implement that? Much appreciated!

**Reply by ben95cd** [reply]
Hey @schlomo this feature would definitely be able to address that use case. The CI pipeline could mint a signed JWT and present it to Keycloak with the `sub` specifying the user you want to impersonate. Keycloak would then respond with an access token for that user. The RFC doesn't specifically cover just in time provisioning of users, however, you could always create the user right before you send the assertion grant request with the Keycloak API.

**Reply by schlomo** [reply]
I also added https://github.com/keycloak/keycloak/discussions/24286 in an attempt to describe where I got stuck with the current Keycloak.

Creating users via API is very much an anti-thesis to the whole motivation for doing this, which is to create a trust relationship between a single git repo and its related CI pipelines and our Keycloak. That means that the OIDC ID token from CircleCI (GitHub Actions works just the same way) is the *only means* of authentication and also authorization here.

Would it be possible to create some kind of mapping rules in Keycloak to control the automatic user generation and to define the scopes or roles or groups that can be granted based on the claims that are encoded in the OIDC token?

This feature is really important to get rid of static credentials for "test users" that get hard-coded or shared around.

**Comment by Waterstraal** [comment]
We have a use case where we want to generate a PDF by visiting a webpage with Puppeteer from within an API. To be able to do this, Puppeteer must be authorized for the given page. We want to reuse the session / token of the user that requests the PDF. 

Will this RFC address such a use case? Or are there alternative methods of solving this issue? Thank you!

**Comment by glennlod** [comment]
Hi,

I have a usecase where I need to be able to support both client credentials as the JWT Bearer Authorization grant. However this is not yet implemented I believe ? JWT Bearer assertion (authentication) does not cover our needs.
Is there any idea on timeline ?

Thank you!

**Comment by stianst** [comment]
Adding my 2-cents to this.

RFC-7521 outlines a few common scenarios:

1. Client authentication
2. Client Acting on Behalf of Itself
3. Client Acting on Behalf of a User
4. Client Acting on Behalf of an Anonymous User

1 and 2 we already support, so making it a bit more clear in this discussion we're talking about client acting on behalf of a user. The anonymous user scenario is basically just 3, but with some special anonymous user.

Simply implementing the spec is one thing, but the harder thing is to support the needed policies in order to control what should be permitted. There's a few parts to this:

a. Ability to configure valid issuers
b. Ability to configure permission on what authorization grants are permitted for a given client (which subjects, audience, etc, can be included in the issues tokens)
c. Mapping subject for assertions to internal users

In terms of configuration, neither a or b should be part of the client configuration itself, but should be externalized.

For trusted issuers we do already have the concept of identity brokers, which could possible be used for this purpose like proposed in the description. However, one key problem there is identity brokers are currently used for federated authentication, and we would at least need some marker on an identity provider if it can also be used as an assertion issuer, where it could be that a particular provider is only an assertion issuer, and can not also be used for delegated authentication (idp redirect flow). Further, we would need the ability to control what tokens can be used, and make sure invalid tokens can't be used (an idp for instance issues tokens to many clients). This probably means the token should be issued to the client, and not anything else, so perhaps the expected audience in the assertion should be configurable on a per-client basis?

With regard to the second part of configuring permissions on what authorization grants can be permitted, we should look towards token exchange for that, as this really is just an alternative "protocol" for an "external to internal" token request.

For the last part on how we map subjects for external tokens into internal subjects, not quite sure how that would look like. If a assertion provider is an identity provider then we could use identity links to do that, otherwise we'd need some alternative mechanism (email?! may need to be configurable).

**Reply by jbman** [reply]
I agree with this summary. 

Just some additional notes and proposals to see if I understand it correctly: 
>In terms of configuration, neither a or b should be part of the client configuration itself, but should be externalized.

Referring from a client to an IdP configuration for a client is an externalization, right? And a client could point to an IdP config disabled for login, however it could still be used to define the trusted issuer.
I think if trusted issuer should not be used as IdP, it would be better to have the configuration not externalized it is just the trusted issuer URI stored at the client.

> Further, we would need the ability to control what tokens can be used, and make sure invalid tokens can't be used (an idp for instance issues tokens to many clients)

The token should be issued to the authorization server (https://datatracker.ietf.org/doc/html/rfc7523#section-3)
In terms of configuration, Keycloak needs a configuration which audience is expected for a trusted IdP issuer. The expected audience could be the token URL by default (as suggested in section 3.3). However IdPs often don't allow to choose arbitrary client IDs (i.e. the token URL or a Keycloak realm), so this expected own identity should be configurable.

The expected audience value would be a second attribute stored at the client (in addition to the issuer or reference to the IdP config to define the trusted issuer).

> With regard to the second part of configuring permissions on what authorization grants can be permitted, we should look towards token exchange for that, ...

I don't see any options or a additional features necessary here when a Keycloak client is used and has enabled "Using JWTs as Authorization Grants" protocol. The Keycloak client scope mechanisms would be applied to create the token with claims like produced by the auth code grant. It is just not possible to get consents granted by the user during the authorization request.

> For the last part on how we map subjects for external tokens into internal subjects, not quite sure how that would look like. I

Yes, there could a SPI to resolve the user fro the the "sub" claim value. Keycloak could provide "Find by IdP Link" and "Find subject as username". 
Also not sure if the user should be resolved by some other IdP claims like "email". This could be left to custom implementations.

**Comment by rouke-broersma** [comment]
This is necessary for federated workload identity which is becoming increasingly relevant for 'passwordless' authentication between (cloud) services. 

This is already implemented by major IdPs such as Microsoft Entra, Google  Cloud, Amazon Cognito, Github Workflows, Azure Devops et cetera. Enabling this in Keycloak would enable Keycloak as a in-cluster identity store for Kubernetes inter-service authentication/authorization but also for passwordless deployment to Kubernetes clusters (and other oidc enabled platforms).

**Reply by slaskawi** [reply]
Related: https://github.com/keycloak/keycloak/issues/37600

**Comment by AxelNennker** [comment]
The [LInux Foundation's CAMARA project](https://camaraproject.org/) is using this grant-type `urn:ietf:params:oauth:grant-type:jwt-bearer`.

The definition of that is in [CAMARA's Security and Interoperability](https://github.com/camaraproject/IdentityAndConsentManagement/blob/main/documentation/CAMARA-Security-Interoperability.md#jwt-bearer-flow) document.
Descriptions of the flow is in [CAMARA APIs Access and User Consent Management](https://github.com/camaraproject/IdentityAndConsentManagement/blob/main/documentation/CAMARA-API-access-and-user-consent.md#jwt-bearer-flow)

A couple of authorization server vendor's working with CAMARA API providers implemented the grant type using keycloak.

Usecase include e.g. [Fraud Prevention and Detection](https://w3c.github.io/dpv/2.0/dpv/modules/purposes.html#FraudPreventionAndDetection), Signup and Account Creation using e.g. he CAMARA API's [NumberVerification](https://github.com/camaraproject/NumberVerification) or [Sim-Swap](https://github.com/camaraproject/SimSwap/).

**Comment by ahus1** [comment]
This is currently been worked on in https://github.com/keycloak/keycloak/issues/43152 and some part of it will in 26.5.

Please follow that feature for more updates.
